### PR TITLE
feature: Add validation for NaN values in datablocks

### LINKED ISSUE #433: Validation error for nans in TimeSeries/T3D/etc missing
**Describe the bug**
When writing a TimeSeries (or ForcingModel) containing nans, there is no validation error. However, the FM kernel is not able to handle nan values and it will crash with some erro. Therefore, nans in the file are not valid FM input, so should be validated by hydrolib-core.

**To Reproduce**
```
from pathlib import Path
import numpy as np
import hydrolib.core.dflowfm as hcdfm

datablock = np.random.uniform(low=-40, high=130.3, size=(4,2))*np.nan
datablock_list = datablock.tolist()

steric = hcdfm.TimeSeries(
    name="east2_0001",
    quantityunitpair=[hcdfm.QuantityUnitPair(quantity="time", unit="seconds since 2022-01-01 00:00:00 +00:00"),
                      hcdfm.QuantityUnitPair(quantity="waterlevel", unit="m")],
    timeinterpolation=hcdfm.TimeInterpolation.linear,
    datablock=datablock_list, 
    )
bc_east = hcdfm.ForcingModel()
bc_east.forcing.append(steric)
bc_east.save(filepath=Path("./steric_east2_hydrolib.bc"))
```

**Expected behavior**
Error when nans in dataset.

**Version info (please complete the following information):**
 - OS: Windows, Spyder
 - Version 0.3.1 (main branch of today)

### PR DESCRIPTION
**Pull request by tim-vd-aardweg**
Fixes #433

### CODE REVIEW

**Comment by tim-vd-aardweg** [comment]
There is one code smell remaining, but I prefer to have the error message inside the tests, rather than as a variable somewhere else.

File: hydrolib/core/dflowfm/ini/models.py [hunk 1]
```diff
@@ -204,6 +207,34 @@ def convert_value(
...
+        Raises:
+            ValueError: When a NaN is present in the datablock.
+
+        Returns:
+            Datablock: The validated datablock.
+        """
+        for list in datablock:
+            for value in list:
+                if cls._is_float_and_nan(value) or cls._is_string_and_nan(value):
+                    raise ValueError("NaN is not supported in datablocks.")
```
Comment (arthurvd) [inline hydrolib/core/dflowfm/ini/models.py:226]: You could do this shorter:
```python
if any(cls._is_float_and_nan(value) for list in datablock for value in list):
    raise ValueError("NaN is not supported in datablocks.")
```
Not sure whether that's more readable. (but SonarCloud always gives me penalty points for deeply nested control structures)

**Review by arthurvd — CHANGES_REQUESTED** [review]
Please see my remarks/suggestions.

File: tests/utils.py [hunk 2]
```diff
@@ -89,3 +87,20 @@ def assert_file_is_same_binary(
         assert filecmp.cmp(input_path, reference_path)
     else:
         assert not input_path.exists()
+
+
+def error_occurs_only_once(error_message: str, full_error: str) -> bool:
+    """Checks if the given error message occurs exactly once in the full error string.
```
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply tests/utils.py]: Good to know! I changed it.

File: tests/utils.py [hunk 3]
```diff
@@ -89,3 +87,20 @@ def assert_file_is_same_binary(
...
+
+    Args:
+        error_message (str): The error to check for.
+        full_error (str): The full error as a string.
+
+    Returns:
+        bool: Returns True if the error message occurs exactly once in the full error.
+        Returns False otherwise.
```
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply tests/utils.py:101]: Done.

File: hydrolib/core/dflowfm/ini/models.py [hunk 4]
```diff
@@ -204,6 +207,34 @@ def convert_value(
...
+
+        return datablock
+
+    @staticmethod
+    def _is_float_and_nan(value: float) -> bool:
+        return isinstance(value, float) and isnan(value)
+
+    @staticmethod
+    def _is_string_and_nan(value: str) -> bool:
```
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply hydrolib/core/dflowfm/ini/models.py:235]: You are correct. I misunderstood what happened here. I assumed (incorrectly) that the datablock would only contain numbers, and that any explicit 'NaN' values would be interpreted as strings. I thought that that was the reason why the `Datablock` contained a `Union[float, str]`. But it seems the datablock can simply contain strings. I will change it and update the tests.

File: hydrolib/core/dflowfm/ini/models.py (hunk 1 above)
Comment (tim-vd-aardweg) [reply hydrolib/core/dflowfm/ini/models.py:226]: The complexity remains the same, but I agree that your suggestion may be the way most people do it in python, so I will use your suggestion! 👍

File: tests/dflowfm/ini/test_models.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+from ...utils import error_occurs_only_once
+
+
+class TestDataBlockINIBasedModel:
+    @pytest.mark.parametrize(
+        "nan_value",
+        [
+            pytest.param(nan),
+            pytest.param("NAN"),
```
[the remark this answers is not in the document]
Comment (tim-vd-aardweg) [reply tests/dflowfm/ini/test_models.py:17]: Done

**Review by arthurvd — APPROVED** [review]

**Pull request merged**